pkg/controller/drivelvgcontroller/drivelvgcontroller.go

```diff
@@ -0,0 +1,365 @@
+const RequeueDriveTime = time.Second * 30
+
+// Controller reconciles drive custom resource
+type Controller struct {
+	client   *k8s.KubeClient
+	crHelper *k8s.CRHelper
+	// CRHelper instance which reads from cache
+	cachedCrHelper *k8s.CRHelper
+	log            *logrus.Entry
+}
+
+// NewDriveController creates new instance of Controller structure
+// Receives an instance of base.KubeClient and logrus logger
+// Returns an instance of Controller
+func NewDriveController(client *k8s.KubeClient, k8sCache k8s.CRReader, log *logrus.Logger) *Controller {
+	return &Controller{
+		client:         client,
+		crHelper:       k8s.NewCRHelper(client, log),
+		cachedCrHelper: k8s.NewCRHelper(client, log).SetReader(k8sCache),
+		log:            log.WithField("component", "Controller"),
+	}
+}
+
+// SetupWithManager registers Controller to ControllerManager
+func (d *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&drivecrd.Drive{}).
+		Watches(&source.Kind{Type: &lvgcrd.LogicalVolumeGroup{}}, &handler.EnqueueRequestForObject{}).
+		WithEventFilter(predicate.Funcs{
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return false
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return d.filterUpdateEvent(e.ObjectOld, e.ObjectNew)
+			},
+		}).
+		Complete(d)
+}
+
+// Reconcile reconciles Drive custom resources
+func (d *Controller) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	defer metricsC.ReconcileDuration.EvaluateDurationForType("csicontroller_drive_controller")()
+	resourceName := req.Name
+	ctx, cancelFn := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancelFn()
+
+	log := d.log.WithFields(logrus.Fields{"method": "Reconcile", "name": resourceName})
+
+	drive := &drivecrd.Drive{}
+	if err := d.client.ReadCR(ctx, resourceName, "", drive); err == nil {
+		return d.reconcileDrive(ctx, drive)
+	}
+	log.Warnf("Failed to read Drive %s CR, try to read LVG CR", resourceName)
+	lvg := &lvgcrd.LogicalVolumeGroup{}
+	if err := d.client.ReadCR(ctx, resourceName, "", lvg); err != nil {
+		log.Errorf("Failed to read LVG %s CR", resourceName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	return d.reconcileLVG(lvg)
+}
+
+// reconcileLVG perform logic for LVG reconciliation
+func (d *Controller) reconcileLVG(lvg *lvgcrd.LogicalVolumeGroup) (ctrl.Result, error) {
+	log := d.log.WithFields(logrus.Fields{
+		"method": "reconcileLVG",
+	})
+	var (
+		status = lvg.Spec.GetStatus()
+		health = lvg.Spec.GetHealth()
+		name   = lvg.GetName()
+	)
+	// If LVG status is failed or Health is not good, try to reset its AC size to 0
+	if status == apiV1.Failed || health != apiV1.HealthGood {
+		return ctrl.Result{}, d.resetACSizeOfLVG(name)
+	}
+	// If LVG is already presented on a machine but doesn't have AC, try to create its AC using annotation with
+	// VG free space
+	size, err := getFreeSpaceFromLVGAnnotation(lvg.Annotations)
```

Comment:
Why do we need to use LVG annotations when original size is known and you can read related volumes info? Is that for SYSLVG?

Reply:
Yes, it's used for system LVG